### PR TITLE
Memory Stream - File Stream (win/lin) - HTTP Request / Response - Updates

### PR DESCRIPTION
**Pull request by vhvb1989**
Adding:
-  File Stream for Linux and Windows.  Customer is responsible for opening the file to get a handle. Then we use that handle to construct the BodyStream for the File.  See the basic example on how to use it on sdk/Samples/azure_core_with_curl_bodyBuffer.cpp  (it is currently only running a File Stream test)

- Memory Stream
   - Removing the memory stream that holds a vector within. We don't need to have such a stream.
   - Keeping only Memory stream that works with a ptr and size
   - The programming model needs to be adapted to avoid using "constructRequest / CreateRequest" methods where the body content is created and expected to be moved to the memory stream. Instead, create the body content in the same method where the request will be sent to keep the body content in stack. (See the change - examples I made to blob_rest_client.hpp)
   
  - Adding static methods for reading memory stream.
```cpp
// Will loop reading from body until buffer is filled up to count bytes OR until the end of body content
static int64_t ReadToCount(Context& context, BodyStream& body, uint8_t* buffer, int64_t count);

// Will loop reading and inserting bytes into an std::vector<> until the end of body content
static std::unique_ptr<std::vector<uint8_t>> ReadToEnd(Context& context, BodyStream& body);
```

- HTTP Request is updated in this PR to avoid owning the body stream. This is a proposal to make Request have a ref BodyStream& so the customer/convenience layer can keep the body stream ownership instead of moving it to the request. Allowing a customer to reuse a unique_ptr<BodyStream>.

>Note: Storage cmake project is currently disabled because it need to be updated on top of Core Changes.  Please have it enabled by updating CMakeLists.txt upon fixing breaking changes.

>Note2: I am still working on adjusting the internal implementation of the libcurl transport to be fully function with this latest changes.  However, the signatures won't change, so please start fixing breaking changes to Storage client

Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/198
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/195
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/194
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/96

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -43,4 +43,4 @@ include(${CMAKE_SOURCE_DIR}/cmake-modules/doxygen_common.cmake)
 # sub-projects
 add_subdirectory(sdk/core/azure-core)
 add_subdirectory(sdk/samples/http_client/curl) # will work only if BUILD_CURL_TRANSPORT=ON
-add_subdirectory(sdk/storage)
+#add_subdirectory(sdk/storage)
```
Comment (vhvb1989) [inline CMakeLists.txt:46]: uncomment this to re-enable storage cmake proj

File: sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp [hunk 2]
```diff
@@ -82,23 +73,67 @@ int main()
...
+  string host("https://httpbin.org/put");
+  cout << "Creating a File request to" << endl << "Host: " << host << endl;
+
+  int fd = open("/home/vagrant/workspace/a", O_RDONLY);
+  auto requestBodyStream = std::make_unique<FileBodyStream>(fd, 0, 10);
+
+  auto body = Http::BodyStream::ReadToEnd(context, *requestBodyStream);
+  cout << body->data() << endl << body->size() << endl;
+
+  close(fd);
```
Comment (vhvb1989) [inline sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp:89]: linux File Stream Example

File: sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp [hunk 3]
```diff
@@ -82,23 +73,67 @@ int main()
...
+      NULL,
+      OPEN_EXISTING,
+      FILE_FLAG_SEQUENTIAL_SCAN,
+      NULL);
+  auto requestBodyStream = std::make_unique<FileBodyStream>(hFile, 20, 200);
+
+  auto body = Http::BodyStream::ReadToEnd(context, *requestBodyStream);
+  cout << body->data() << endl << body->size() << endl;
+
+  CloseHandle(hFile);
```
Comment (vhvb1989) [inline sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp:116]: Windows File Stream Example

File: sdk/storage/inc/common/memory_stream.hpp [hunk 4]
```diff
@@ -32,7 +30,7 @@ namespace Azure { namespace Storage {
     std::size_t m_offset = 0;
   };
 
-  inline std::unique_ptr<MemoryStream> CreateMemoryStream(const uint8_t* data, std::size_t length)
+  inline std::unique_ptr<MemoryStream> CreateMemoryStream(const uint8_t* data, int64_t length)
   {
     return std::make_unique<MemoryStream>(data, length);
```
Comment (vhvb1989) [inline sdk/storage/inc/common/memory_stream.hpp:35]: Can we remove this MemoryStream ?   Use only the one from Core now?

File: sdk/storage/src/blobs/block_blob_client.cpp [hunk 5]
```diff
@@ -58,11 +58,11 @@ namespace Azure { namespace Storage { namespace Blobs {
   }
 
   BlobContentInfo BlockBlobClient::Upload(
-      std::unique_ptr<Azure::Core::Http::BodyStream> content,
+      Azure::Core::Http::BodyStream& content,
       const UploadBlobOptions& options) const
   {
     BlobRestClient::BlockBlob::UploadOptions protocolLayerOptions;
-    protocolLayerOptions.BodyStream = std::move(content);
+    // protocolLayerOptions.BodyStream = std::move(content);
```
Comment (vhvb1989) [inline sdk/storage/src/blobs/block_blob_client.cpp:65]: We also want to move body content out of options.
BodyStream is a required parameter for HttpRequest, just like url.
So it should not be in operation options.

Confirmed by @jeff  and @RickWinter

File: sdk/storage/test/test_base.hpp [hunk 6]
```diff
@@ -29,9 +29,14 @@ namespace Azure { namespace Storage { namespace Test {
...
-  std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>& stream);
+  inline std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>& stream)
+  {
+    Azure::Core::Context context;
+    return *Azure::Core::Http::BodyStream::ReadToEnd(context, *stream);
+  }
 
-  inline std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>&& stream)
+  inline std::vector<uint8_t> ReadBodyStream(
+      std::unique_ptr<Azure::Core::Http::BodyStream>&& stream)
```
Comment (vhvb1989) [inline sdk/storage/test/test_base.hpp:39]: We would not need these ones anymore.  Update to use the BodyStream::ReadToCount(...) or BodyStream::ReadToEnd(..) functions

File: CMakeLists.txt (hunk 1 above)
Comment (JinmingHu-MSFT) [reply CMakeLists.txt:46]: I think it's a very good practice to disable storage build when you introduce a breaking change in Azure Core.

Then why did you also change storage code?

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    // return copied size
+    virtual int64_t Read(Context& context, uint8_t* buffer, int64_t count) = 0;
+
+    // Closes the stream; typically called after all data read or if an error occurs.
+    virtual void Close() = 0;
+
+    // Keep reading until buffer is all fill out of the end of stream content is reached
+    static int64_t ReadToCount(Context& context, BodyStream& body, uint8_t* buffer, int64_t count)
+    {
+      int64_t readBytes;
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:52]: `readBytes` isn't used out of the for loop, so maybe you want to move it into the loop

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    const uint8_t* m_data;
+    int64_t m_length;
+    int64_t m_offset = 0;
+
+  public:
+    // Forbid constructor for rval so we don't end up storing dangling ptr
+    MemoryBodyStream(std::vector<uint8_t> const&&) = delete;
+
+    MemoryBodyStream(std::vector<uint8_t> const& buffer)
+        : MemoryBodyStream(buffer.data(), static_cast<uint64_t>(buffer.size()))
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:98]: you want to static_cast to int64_t? looks like a typo

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+    int64_t Length() const override { return this->m_length; }
+
+    int64_t Read(Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
+      // Copy what's left or just the count
+      std::memcpy(buffer, this->m_data + m_offset, (size_t)copy_length);
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:115]: I remember compiler will give a warning for `std::memcpy` on Windows.

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    {
+    }
+
+    int64_t Length() const override { return this->m_length; }
+
+    int64_t Read(Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:113]: One conclusion from Jun. 10 meeting is

* Use C++ style casts
* Avoid C casts.

reference to the recording at around 52 min.

Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/inc/http/bodyStream.hpp:113]: ditto for all other casts, since I've seen these C style casts a lot in your code.

File: sdk/core/azure-core/inc/http/http.hpp [hunk 11]
```diff
@@ -215,7 +215,7 @@ namespace Azure { namespace Core { namespace Http {
     std::map<std::string, std::string> m_retryHeaders;
     std::map<std::string, std::string> m_retryQueryParameters;
     // Work only with streams
-    std::unique_ptr<BodyStream> m_bodyStream;
+    BodyStream& m_bodyStream;
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/http.hpp:218]: Why is this changed to reference again? It was just changed to unique_ptr several days ago.

I vaguely remember that we don't allow reference in member variables.

Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/inc/http/http.hpp:218]: here https://azure.github.io/azure-sdk/cpp_design.html#const-and-reference-members

File: sdk/storage/inc/common/memory_stream.hpp (hunk 4 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/common/memory_stream.hpp:35]: Yes, I see you've changed the implementation of `Azure::Core::MemoryStream`, it doesn't make a copy now.

File: sdk/storage/test/test_base.hpp (hunk 6 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/test/test_base.hpp:39]: I think so

File: sdk/core/azure-core/CMakeLists.txt [hunk 12]
```diff
@@ -14,6 +14,13 @@ if(NOT CURL_FOUND)
   find_package(CURL ${CURL_MIN_REQUIRED_VERSION} REQUIRED)
 endif()
 
+if (MSVC)
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/CMakeLists.txt:17]: What if Cygwin and MinGW?

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+          std::min(count, this->m_length - this->m_offset),
+          this->m_baseOffset + this->m_offset);
+      this->m_offset += result;
+      return result;
+    }
+
+    int64_t Length() const override { return this->m_length; };
+
+    // close does nothing opp
+    void Close() {}
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:186]:
```suggestion
    void Close() override {}
```

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+      (void)result;
+
+      this->m_offset += numberOfBytesRead;
+      return numberOfBytesRead;
+    }
+
+    int64_t Length() const override { return this->m_length; };
+
+    // close does nothing opp
+    void Close() {}
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:237]:
```suggestion
    void Close() override {}
```

File: sdk/core/azure-core/CMakeLists.txt [hunk 15]
```diff
@@ -14,6 +14,13 @@ if(NOT CURL_FOUND)
...
+  add_compile_definitions(Windows)
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/CMakeLists.txt:18]: It's uncommon to define macro with lowercase letters?

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    ~FileBodyStream() override {}
+
+    // Rewind seek back to 0
+    void Rewind() override { this->m_offset = 0; }
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      auto result = pread(
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:174]: `pread()` may return `-1` on error.

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      DWORD numberOfBytesRead;
+      auto o = OVERLAPPED();
+      o.Offset = (DWORD)(this->m_baseOffset + this->m_offset);
+      o.OffsetHigh = (DWORD)((this->m_baseOffset + this->m_offset) >> 32);
+
+      auto result = ReadFile(
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:220]: error handling

File: sdk/core/azure-core/inc/http/bodyStream.hpp [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+
+#ifdef Posix
+#include <unistd.h>
+#endif
+
+#ifdef Windows
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <Windows.h>
+#endif // Windows
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/bodyStream.hpp:14]: Nit: is it a good idea to include `unistd.h` and `Windows.h` in header file? It'll pollute the global namespace.

File: sdk/core/azure-core/inc/http/bodyStream.hpp (hunk 16 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/bodyStream.hpp:174]: https://github.com/Azure/azure-sdk-for-cpp/issues/237

File: sdk/core/azure-core/inc/http/bodyStream.hpp (hunk 17 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/bodyStream.hpp:220]: https://github.com/Azure/azure-sdk-for-cpp/issues/237

File: sdk/core/azure-core/inc/http/http.hpp (hunk 11 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/http.hpp:218]: https://github.com/Azure/azure-sdk-for-cpp/issues/238

**Comment by JinmingHu-MSFT** [comment]
@vhvb1989 BTW, are the remaining two bugs in https://github.com/Azure/azure-sdk-for-cpp/issues/198 also fixed in this PR? Because they are still blockers for storage team.

**Comment by vhvb1989** [comment]
> @vhvb1989 BTW, are the remaining two bugs in #198 also fixed in this PR? Because they are still blockers for storage team.

are they the 9 and 10?   Added info about them.
Let me know if there is something else?

I am seeing a seg fault right now, I will fix it here

**Comment by JinmingHu-MSFT** [comment]
> if there is something

9 and 11, I struckthrough the ones that have been fixed.

**Comment by JinmingHu-MSFT** [comment]
> I am seeing a seg fault right now, I will fix it here

You don't need to fix crashes in storage code, we can take care of that. You only need to ensure there's no crash in core code.

**Comment by vhvb1989** [comment]
@JinmingHu-MSFT , yes, 9 and 10 are fixed.

And the seg fault was an error in the sample test cases. I have fixed.

I was getting the hang while reading and found the root cause. I have tested Core transport implementation with the samples I have and it is working fine with the changes on this PR.

Please fix Storage Code whenever possible

About the BodyStream Reference.  Core team will come to a final decision about this tomorrow.  (I will update PR after that)

File: cmake-modules/DefinePlatform.cmake [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+# APPLE  : is TRUE on Apple systems. Note this does not imply the
+#          system is Mac OS X, only that APPLE is #defined in C/C++
+#          header files.
+
+if (WIN32 OR MINGW OR MSYS OR CYGWIN)
+  add_compile_definitions(WINDOWS)
+endif ()
+if (UNIX)
+  add_compile_definitions(POSIX)
+endif ()
```
Comment (JinmingHu-MSFT) [inline cmake-modules/DefinePlatform.cmake:18]: Do we really need to make it so compilcated?

I think in the code

```
#ifdef _WIN32
#include <Windows.h>
#else
#include <unistd.h>
#endif
```
should be enough.

File: sdk/core/azure-core/inc/http/http.hpp [hunk 20]
```diff
@@ -233,14 +233,17 @@ namespace Azure { namespace Core { namespace Http {
...
+    Request(HttpMethod httpMethod, std::string const& url, BodyStream& bodyStream)
+        : m_method(std::move(httpMethod)), m_url(url), m_bodyStream(bodyStream),
           m_retryModeEnabled(false)
     {
     }
 
     // Typically used for GET with no request body.
-    Request(HttpMethod httpMethod, std::string const& url) : Request(httpMethod, url, nullptr) {}
+    Request(HttpMethod httpMethod, std::string const& url)
+        : Request(httpMethod, url, *std::make_unique<NullBodyStream>())
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/http/http.hpp:244]: I don't think this will work, the `std::unique_ptr` will be destructed very soon, then you're referencing a deleted object.

File: sdk/core/azure-core/CMakeLists.txt [hunk 21]
```diff
@@ -22,7 +22,7 @@ add_library (
   src/http/policy.cpp
   src/http/request.cpp
   src/http/response.cpp
-  src/http/stream.cpp
+  src/http/bodyStream.cpp
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/CMakeLists.txt:25]:
```suggestion
  src/http/body_stream.cpp
```

Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/CMakeLists.txt:25]: we have file naming rules. Other files are using snake_case.

File: sdk/storage/inc/common/memory_stream.hpp (hunk 4 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/common/memory_stream.hpp]: done

File: sdk/storage/test/test_base.hpp (hunk 6 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/test/test_base.hpp]: > @JinmingHu-MSFT , yes, 9 and 10 are fixed.
> 
> And the seg fault was an error in the sample test cases. I have fixed.
> 
> I was getting the hang while reading and found the root cause. I have tested Core transport implementation with the samples I have and it is working fine with the changes on this PR.
> 
> Please fix Storage Code whenever possible
> 
> About the BodyStream Reference. Core team will come to a final decision about this tomorrow. (I will update PR after that)

9 isn't fixed.

Here https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/core/azure-core/src/http/curl/curl.cpp#L444,

if I remove Line 444, then 9 is fixed.

I didn't look into it carefully, so please double check the logic here is correct.

File: sdk/core/azure-core/inc/http/http.hpp (hunk 20 above)
Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/inc/http/http.hpp:244]: This leads to a crash issue.

**Comment by JinmingHu-MSFT** [comment]
@vhvb1989 I've fixed everything in Storage Blobs Service, @katmsft will fix DataLake service. However storage unittests still cannot pass. Some of the bugs in core code haven't been fixed yet, and some new ones are introduced in this PR.

@katmsft I disabled datalake build. https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/CMakeLists.txt#L78 https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/CMakeLists.txt#L83  https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/sample/CMakeLists.txt#L11

File: sdk/storage/src/blobs/block_blob_client.cpp (hunk 5 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/src/blobs/block_blob_client.cpp]: Blob service changed.

Comment (katmsft) [reply sdk/storage/src/blobs/block_blob_client.cpp]: I am still in doubt why body stream should be required for **ALL** requests. A request body should never be mandatory even for PUT requests.

Comment (JinmingHu-MSFT) [reply sdk/storage/src/blobs/block_blob_client.cpp]: I agree that request body shouldn't be mandatory for GET/HEAD/DELETE requests. But I think it's mandatory for PUT/POST requests.

File: sdk/core/azure-core/inc/http/http.hpp (hunk 11 above)
Comment (katmsft) [reply sdk/core/azure-core/inc/http/http.hpp:218]: It would be best for us to avoid this kind of breaking changes in the future. Changing from A to B is fine, but changing from A to B to C then back to A seems to be badly planned or designed.
Also, there is no detailed information in the linked issue for the Storage team to know the reasoning behind this change.

File: sdk/core/azure-core/src/http/bodyStream.cpp [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+    if (totalRead == count || readBytes == 0)
+    {
+      return totalRead;
+    }
+  }
+}
+
+std::unique_ptr<std::vector<uint8_t>> BodyStream::ReadToEnd(Context& context, BodyStream& body)
+{
+  constexpr int64_t chunkSize = 1024 * 8;
```
Comment (katmsft) [inline sdk/core/azure-core/src/http/bodyStream.cpp:46]: Magic numbers should be documented. Ideally, this is a changeable size for the client.

Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/src/http/bodyStream.cpp:46]: @katmsft Let's be less picky about these trivial stuff and focus on the ability to function properly at this stage.

File: sdk/storage/src/blobs/block_blob_client.cpp (hunk 5 above)
Comment (JeffreyRichter) [reply sdk/storage/src/blobs/block_blob_client.cpp]: HTTP allows request bodies for ALL HTTP methods and so the request supports this. Also, the request object doesn't even know the exact method that's being used (it's passed as a parameter). So, since every method CAN have a body and because the body is a separate logical entity of any arbitrary HTTP request the body is an explicitly specified parameter. We added a NullBodyStream which can be passed for HTTP requests that want to have no body.

File: sdk/core/azure-core/inc/http/http.hpp (hunk 11 above)
Comment (JeffreyRichter) [reply sdk/core/azure-core/inc/http/http.hpp:218]: We're experimenting with different designs and getting input from Billy/Casey. Right now, I cannot guarantee that this will not change again - in fact, we're meeting on it again today. We want to get this right and there are nuances. C++ makes it very difficult; in other languages this was trivial. Please just bear with us a while longer.

File: sdk/core/azure-core/inc/http/bodyStream.hpp (hunk 18 above)
Comment (JeffreyRichter) [reply sdk/core/azure-core/inc/http/bodyStream.hpp]: Obviously, we prefer not to do this. For linux we can problem get rid of it because a file descriptor is an 'int'.
For Windows, the HANDLE is the data type. Maybe we can get rid of this header file with a small hack and define the FileBodyStream ctor parameter as a uint32_t. We'll just have to add some comment/documentation.

File: sdk/core/azure-core/inc/http/http.hpp (hunk 20 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/http.hpp]: fixed.

File: sdk/storage/test/test_base.hpp (hunk 6 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/test/test_base.hpp]: Now it's fixed

**Comment by JinmingHu-MSFT** [comment]
@vhvb1989 There's only 1 bug left unfixed in https://github.com/Azure/azure-sdk-for-cpp/issues/198. After fixing that one, blob unit tests can pass.

Although I did see some flaky crash, but i'm not going to make big deal out of it. We can fix that later.

**Comment by vhvb1989** [comment]
storage unit tests passing:
image

@JinmingHu-MSFT , please approve changes to merge this PR.

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp [hunk 23]
```diff
@@ -1063,8 +1063,7 @@ namespace Azure { namespace Storage { namespace Blobs {
           std::vector<uint8_t> bodyContent;
           if (bodyStream->Length() == -1)
```
Comment (katmsft) [inline sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1064]: What does this mean?

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1064]: unknown length, it's Azure Core's design.
